Fasta/q parser refactoring

eclarke opened the pull request:
Hi @johanneskoester, this is my take on a limited refactoring of the fasta and fastq parsers based on issue #133. I'm still new to Rust so I'm very open to critiques here. There are no breaking changes to the API.

What's changed:

1. Record structs lose `header` field: `id` and `desc` are now parsed during read. I know in #52 you mentioned this was done so that you could skip a record without parsing. In my informal tests it looks like there is an extremely small performance penalty to just doing the parsing anyway (on a 900k record Fastq file, there was a 0.01s slowdown on my laptop).
2. Record fields are now public (this is the only API change).
3. There is a new constructor for records: `Record::from_attrs(...)` that builds a record using the same argument signatures as the `Writer::write(...)` method. I'm not in love with the constructor name but I don't really know the conventions in Rust.

johanneskoester commented:
Nice! You can call the constructor `Record::new()`.

eclarke commented:
Currently `Record::new()` creates an empty record, which is used in the `read()` methods, so I didn't want to change that.

johanneskoester commented on src/io/fastq.rs:
```diff
@@ -82,25 +96,41 @@ impl<R: io::Read> Reader<R> {
-    seq: String,
-    qual: String,
+    pub id: String,
+    pub desc: String,
```
desc should be `Option<String>`.

johanneskoester replied:
I'd prefer these attributes to not be public. Is there a specific reason for making them public?

johanneskoester commented on src/io/fasta.rs:
```diff
@@ -476,23 +481,38 @@ impl<W: io::Write> Writer<W> {
-    seq: String,
+    pub id: String,
+    pub desc: String,
+    pub seq: String,
```
As below for fastq: desc should be an option type, and preferably the attributes should not be public.

eclarke commented:
Agree that the fields don't need to be public. Implemented the requested changes, and renamed the constructor from `from_attrs` to `with_attrs`, which seems more canonical.

johanneskoester commented on src/io/fasta.rs:
```diff
@@ -70,7 +70,16 @@ impl<R: io::Read> Reader<R> {
         }
-        record.header.push_str(&self.line);
+        record.id = match self.line[1..].trim_right().splitn(2, ' ').nth(0) {
+            Some("") => "".to_owned(),
```
This line is superfluous, isn't it? It is covered by the id case below. Further, it would be better to use Option::map here.

eclarke replied:
You're right, it was redundant. I'm not sure I understand how to handle the None option with map() though. Wouldn't it be something like 
```rust
...nth(0).map(|s| s.to_owned())
```
and if so, what would happen in the None case?

johanneskoester replied:
Nothing. None will just be passed through.

johanneskoester commented on src/io/fasta.rs:
```diff
@@ -509,15 +533,18 @@ impl Record {
-            Some("") => None,
-            value => value,
+        match self.id.as_ref() {
+            "" => None,
```
I know this breaks the previous API, but I think id should always return a string. There cannot be a fastq/fasta record without an id. If there is no letter behind the separator, it is simply the empty id. What do you think?

eclarke replied:
I like returning a string rather than an option. I've modified it in both parsers to do that.

johanneskoester commented:
Thanks! We are close. I just noticed some other minor issues.

eclarke commented:
Changes made, and learned a bit about splitn and map! Thanks for your patience.

johanneskoester commented:
Great work, thanks!
